File: android-app/app/src/main/java/com/healthdecoder/app/local/DeviceIdentity.kt
package com.healthdecoder.app.local

import android.content.Context
import com.google.gson.JsonObject
import com.healthdecoder.app.network.NetworkModule
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * Anonymous per-install identity for the AI proxy (see BackendAiClient). Phone OTP sign-in is
 * optional/off by default, so most phones never become a logged-in user — this registers a
 * UUID the app generates once with the backend (no SMS, no login) and gets back a long-lived
 * device token, so /api/ai/generate can still meter/pool a Gemini key per install.
 */
object DeviceIdentity {

    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()
    }

    /**
     * Returns a usable device token, registering with the backend first if none is cached yet.
     * Synchronous/blocking (matches GeminiClient/BackendAiClient's style) — call only from a
     * background thread. Returns null if registration fails (e.g. no connectivity); callers
     * should surface a clear "can't reach the server" error rather than falling back to any
     * embedded key.
     */
    fun ensureToken(context: Context): String? {
        AppSettings.getDeviceToken(context)?.let { return it }

        val deviceId = AppSettings.getOrCreateInstallId(context)
        val body = JsonObject().apply { addProperty("deviceId", deviceId) }

        // Same host fallback as BackendAiClient.generate(): the Function URL sits on a different
        // domain from the rest of the backend, and a network that blocks it would leave the app
        // with no device token at all — which reads to the user as "every scan fails" even though
        // the identical handler is reachable through API Gateway.
        val hosts = listOf(NetworkModule.AI_PROXY_BASE_URL, NetworkModule.resolveBaseUrl(context)).distinct()

        for (host in hosts) {
            val request = Request.Builder()
                .url("${host}api/device/register")
                .addHeader("ngrok-skip-browser-warning", "true")
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .build()
            try {
                client.newCall(request).execute().use { response ->
                    val text = response.body?.string().orEmpty()
                    if (response.isSuccessful) {
                        val token = com.google.gson.JsonParser.parseString(text)
                            .asJsonObject.get("token")?.asString?.takeIf { it.isNotBlank() }
                        if (token != null) {
                            AppSettings.setDeviceToken(context, token)
                            return token
                        }
                    }
                }
            } catch (e: Exception) {
                // Transport failure against this host — try the next one.
            }
        }
        return null
    }
}

File: ios-app/HealthDecoder/Local/DeviceIdentity.swift
import Foundation

/// Anonymous per-install identity for the AI proxy (see `BackendAiClient`). Phone OTP sign-in
/// is optional/off by default, so most installs never become a logged-in user — this registers
/// a UUID the app generates once with the backend (no SMS, no login) and gets back a
/// long-lived device token, so `/api/ai/generate` can still meter/pool a Gemini key per install.
///
/// Direct port of `local/DeviceIdentity.kt`. Kotlin's version is synchronous/blocking (called
/// off the main thread, matching GeminiClient/BackendAiClient's old style); this is an `actor`
/// instead so registration is safely async/await, and concurrent callers awaiting `ensureToken()`
/// while a registration is already in flight share that one request rather than each firing
/// their own `POST /api/device/register`.
actor DeviceIdentity {
    static let shared = DeviceIdentity()

    private init() {}

    private let session = URLSession(configuration: {
        let c = URLSessionConfiguration.default
        c.timeoutIntervalForRequest = 30
        c.timeoutIntervalForResource = 30
        return c
    }())

    private var inFlight: Task<String?, Never>?

    /// Returns a usable device token, registering with the backend first if none is cached yet.
    /// Returns nil if registration fails (e.g. no connectivity) — callers must surface a clear
    /// "can't reach the server" error rather than falling back to any embedded key.
    func ensureToken() async -> String? {
        if let cached = AppSettings.deviceToken, !cached.isEmpty { return cached }

        if let inFlight { return await inFlight.value }

        let task = Task<String?, Never> { [weak self] in
            await self?.register()
        }
        inFlight = task
        let result = await task.value
        inFlight = nil
        return result
    }

    private func register() async -> String? {
        let deviceId = AppSettings.getOrCreateInstallId()
        guard let url = URL(string: "\(APIClient.aiProxyBaseURL)api/device/register") else { return nil }
        guard let bodyData = try? JSONSerialization.data(withJSONObject: ["deviceId": deviceId]) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("true", forHTTPHeaderField: "ngrok-skip-browser-warning")
        request.httpBody = bodyData

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return nil }
            guard
                let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                let token = (root["token"] as? String)?.trimmingCharacters(in: .whitespaces), !token.isEmpty
            else { return nil }
            AppSettings.deviceToken = token
            return token
        } catch {
            return nil
        }
    }
}
